### PR TITLE
Add homework 5

### PR DESCRIPTION
**Pull request by LakustaL**

### CODE REVIEW

File: students/Lera_Lakusta/05-bag-of-words/conclusion.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+1. Скрейпінг сайту, як Розетка, - це довго, тому що ~ кожні 10к коментарів блокують.
```
[the remark this answers is not in the document]
Comment (LakustaL) [reply students/Lera_Lakusta/05-bag-of-words/conclusion.md:1]: Так, я додавала sleep, але мене все рівно вибивало... Сам сайт не блокував, але просто відхиляло запит у певний момент.

File: students/Lera_Lakusta/05-bag-of-words/BoW.ipynb [hunk 2]
```diff
@@ -0,0 +1,795 @@
...
+     },
+     "execution_count": 7,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "def choose_label (label):\n",
+    "    if label:\n",
+    "        if 5 >= label >= 4:\n",
```
[the remark this answers is not in the document]
Comment (LakustaL) [reply students/Lera_Lakusta/05-bag-of-words/BoW.ipynb:309]: Дякую!

File: students/Lera_Lakusta/05-bag-of-words/BoW.ipynb [hunk 3]
```diff
@@ -0,0 +1,795 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "vectorizer = CountVectorizer()\n",
+    "train_vec = vectorizer.fit_transform(X_train).toarray()\n",
```
[the remark this answers is not in the document]
Comment (LakustaL) [reply students/Lera_Lakusta/05-bag-of-words/BoW.ipynb:366]: Зрозуміла, дякую

File: students/Lera_Lakusta/05-bag-of-words/BoW.ipynb [hunk 4]
```diff
@@ -0,0 +1,795 @@
...
+    "            lemma = morph.parse(token)[0].normal_form\n",
+    "            sent_lemmas.append(lemma)\n",
+    "        sent = \" \".join(sent_lemmas)\n",
+    "        doc_lemmas.append(sent)\n",
+    "    return doc_lemmas\n",
+    "\n",
+    "X_train_lemmas = get_lemmas(X_train)\n",
+    "X_test_lemmas = get_lemmas(X_test)\n",
+    "        \n",
+    "vectorizer = CountVectorizer(analyzer = 'char', ngram_range = (1,5))\n",
```
[the remark this answers is not in the document]
Comment (LakustaL) [reply students/Lera_Lakusta/05-bag-of-words/BoW.ipynb:449]: Я робила декілька спроб з різними типами нграм (символи, слова) і підбирала різну розмірність. Така комбінація виявилася найкращою. А не враховувати пробіли можна так: `tokenizer=lambda x: x.split(' ') `?

[the remark this answers is not in the document]
Comment (mariana-scorp) [reply students/Lera_Lakusta/05-bag-of-words/BoW.ipynb:449]: Ні, `tokenizer` можна додати, тільки якщо `analyzer = 'word'`. Щоб оминути символьних енграм між словами, можна використати `analyzer = 'char_wb'` або побудувати їх вручну.

[the remark this answers is not in the document]
Comment (LakustaL) [reply students/Lera_Lakusta/05-bag-of-words/BoW.ipynb:449]: Ясно, дякую, буду знати)

**Comment by vseloved** [comment]
Загалом, все добре. Рішення щодо нейтрального класу, в принципі, допустиме, і воно себе виправдовує.
З того, що краще було б зробити по-іншому — це upsampling для навчання замість downsampling'у. А також у якості оціночної треба використовувати все-таки вибірку з розподілом, що відповідає реальному. У будь-якому разі, є weighted f1, який рахує якість ніби вона вирівняна. Але accuracy та micro f1 буде відрізнятись.
Що стосується якості, то покращення на 4 пункти — це все-таки 20% помилки :) Подальших покращен можна досягти з використанням більш просунутих моделей та підбору гуперпараметрів. Ну і, звісно, фільтрація даних та додавання цікавих ознак типу не_<слово>...
Оцігка: 95
